### DISCUSSION TITLE
Does PGM take csv or excel files?

### DISCUSSION THREAD

**Discussion by PowerIsKnowledgeOrSomething** [PowerGridModel/.github#29] [Q&A] [answered]
Hello! I'm new to PGM, and I am wondering if PGM works with CSV or Excel files?

**Comment by mgovers** [comment, marked as answer]
Hi and welcome to our community!

The `power-grid-model` library does not directly support those data formats. However, we do have a `power-grid-model-io` library with conversions to the PGM native data type. That library provides a [Tabular converter](https://power-grid-model-io.readthedocs.io/en/stable/converters/tabular_converter.html) for tabular data formats like CSV or Excel.

You will need a mapping file that contains the conversion strategy for each column. The [documentation](https://power-grid-model-io.readthedocs.io/en/stable/converters/tabular_converter.html) will explain to you how to write one.

If, however, you are looking for any data format you can use, we recommend using the PGM-native JSON schema or msgpack schema described in the documentation on [serialization](https://power-grid-model.readthedocs.io/en/stable/user_manual/serialization.html).

I hope this answers your question. Please let me know if you have any additional questions.

Martijn

**Reply by PowerIsKnowledgeOrSomething** [reply]
Ok! Thank you very much. I need to convert Power Factory DGS files into CSV or Excel files so I can have PGM read them.

**Reply by mgovers** [reply]
Power Factory conversions are not provided by us yet. We do have https://github.com/PowerGridModel/power-grid-model-io/issues/123 as an open issue, but it is not prioritized yet. In the meantime, we indeed recommend exporting to CSV/Excel and then use the Tabular converter of `power-grid-model-io` in combination with your own custom mapping file.

If you are willing, you can contribute that mapping file to us as well, so that we can make this workflow easier for other people as well.

**Reply by PowerIsKnowledgeOrSomething** [reply]
Yea the way to do it seems to be to write a script in Power Factory to convert the DGS files into CSV or Excel files and then write a python script in PGM to read the data from those files into PGM. Does that make sense? It's a Final Year Project for my Electrical Engineering degree and I've signed Confidentiality agreements with an Electrical Distribution Company, but I'll pass on the mapping file if I can!

**Reply by mgovers** [reply]
It is good to know that this is a longer project. I will provide my input below.

## Using CSV as an intermediate format

### Quick-and-dirty version

For a first version, I think it makes sense to create a separate script that does both of the things, rather than a separate script for each step.

### Sustainable solution using CSV conversion

As your project grows, the combined script then can be changed to a `main` file that depends on separate libraries for each of the steps. This would probably look something like

```py
intermediate_data = export_dgs_to_power_factory_csv(input_file)
pgm_data = convert_power_factory_csv_to_pgm(intermediate_data)
```

the `convert_power_factory_csv_to_pgm` function in the above can then make use of the `TabularConverter` in `power-grid-model-io`. In that case, the mapping file could be contributed, or the entire solution in an example if you are allowed and willing to do so.

## Sustainable solution with direct conversion

It is actually a really good idea to go from DGS directly to the PGM data format. This, however, is likely a slightly bigger project, as it would involve a full programming solution instead of 'just' a mapping file. That said, depending on the size and potential other parts of your final year project, this could definitely be a possible path going forward as well. The [pandapower converter](https://power-grid-model-io.readthedocs.io/en/stable/converters/pandapower_converter.html) is an example of a similar internship project. In that case, we are definitely open to include your entire contribution directly in the `power-grid-model-io` library if you are allowed and willing to do so.

### Rationale

The PGM-IO repository by nature contains a library. Full library solutions like direct conversion from PowerFactory to PGM are encouraged.

Because of its library nature, we deliberately do not put scripts into PGM-IO. However, you can create and contribute an example notebook in the [`examples` directory](https://github.com/PowerGridModel/power-grid-model-io/tree/main/docs/examples).

Because it concerns conversions, the PGM repository would not be the right location in the first place.

**Reply by PowerIsKnowledgeOrSomething** [reply]
Ok I'll give that a go, thanks!

**Reply by PowerIsKnowledgeOrSomething** [reply]
Hey! A few questions about PGM,
How many ways can PGM represent distributed electrical resources, especially rooftop solar, and solar + battery?​ Can PGM represent on-load tap changers in the distribution transformer?​ And can PGM represent LV voltage regulators/solid state?
Any help would be much appreciated.

**Reply by mgovers** [reply]
Hi @PowerIsKnowledgeOrSomething,

> How many ways can PGM represent distributed electrical resources, especially rooftop solar, and solar + battery?​

PGM does not support solar panel/battery/... directly. However, it does have a component called [`sym_gen`](https://power-grid-model.readthedocs.io/en/stable/user_manual/components.html#load-generator-concrete-types) (or [`asym_gen`](https://power-grid-model.readthedocs.io/en/stable/user_manual/components.html#load-generator-concrete-types) for 3-phase components). The [`LoadGenType`](https://power-grid-model.readthedocs.io/en/stable/api_reference/python-api-reference.html#power_grid_model.enum.LoadGenType) can then be used to set the response to voltage change via the [`type`](https://power-grid-model.readthedocs.io/en/stable/user_manual/components.html#generic-load-and-generator) attribute. To model most real-world generator types, you can create a combination of the three generator types (effectively implementing the ZIP model)

For more complex voltage dependencies than just `const_impedance`, `const_power` and `const_current`, we have an open issue in https://github.com/PowerGridModel/power-grid-model/issues/169 that has not been prioritized yet. If this is something you really need, feel free to upvote that ticket and/or comment on it.

> Can PGM represent on-load tap changers in the distribution transformer?

This is a feature that we are actively working on in https://github.com/PowerGridModel/power-grid-model/issues/74 . We expect to deliver this as an experimental feature in the upcoming weeks and make it generally available to the public in the coming few months. Since it is a big feature, we will do a minor version bump and announce it via our communication channels like the [mailing list](https://lists.lfenergy.org/g/powergridmodel).

> And can PGM represent LV voltage regulators/solid state?

We do not support this as of now. This would require transformers with a decimal tap position, which is not yet prioritized. It would probably need something like automatic tap adjustment calculations using an iterative method, which we have talked about before, but do not have a ticket for at the moment. If you need it, please create a new issue on the issue board of the PGM repo.

As a side-note, please don't forget to register for our hybrid [Power Grid Model Meet-up](https://community.linuxfoundation.org/events/details/lfhq-lf-energy-presents-5th-power-grid-model-meet-up/) via https://community.linuxfoundation.org/events/details/lfhq-lf-energy-presents-5th-power-grid-model-meet-up/ . It will be a great opportunity to see how other people use the power grid model, as well as network with them. it will also be a good opportunity to provide input for new features or other feedback.

**Reply by PowerIsKnowledgeOrSomething** [reply]
I see, Thanks!

**Reply by mgovers** [reply]
FYI: `power-grid-model` version [`v1.8.0`](https://github.com/PowerGridModel/power-grid-model/releases/tag/v1.8.0) was released yesterday, which, among other things, contains support for power flow calculations with automatic tap adjustment.
